easybuild/tools/filetools.py

```diff
@@ -460,20 +467,32 @@ def download_file(filename, url, path, forced=False):
     attempt_cnt = 0
 
     # use custom HTTP header
-    url_req = urllib2.Request(url, headers={'User-Agent': 'EasyBuild',  "Accept" : "*/*"})
+    headers = {'User-Agent': 'EasyBuild', 'Accept': '*/*'}
+    if not HAVE_REQUESTS:
+        url_req = urllib2.Request(url, headers=headers)
 
     while not downloaded and attempt_cnt < max_attempts:
         try:
-            # urllib2 does the right thing for http proxy setups, urllib does not!
-            url_fd = urllib2.urlopen(url_req, timeout=timeout)
-            _log.debug('response code for given url %s: %s' % (url, url_fd.getcode()))
+            if HAVE_REQUESTS:
+                url_req = requests.get(url, headers=headers, stream=True, timeout=timeout)
+                status_code = url_req.status_code
+                url_req.raise_for_status()
+                url_fd = url_req.raw
+                url_fd.decode_content = True
+            else:
+                # urllib2 does the right thing for http proxy setups, urllib does not!
+                url_fd = urllib2.urlopen(url_req, timeout=timeout)
+                status_code = url_fd.getcode()
+            _log.debug('response code for given url %s: %s' % (url, status_code))
             write_file(path, url_fd.read(), forced=forced, backup=True)
             _log.info("Downloaded file %s from url %s to %s" % (filename, url, path))
             downloaded = True
             url_fd.close()
-        except urllib2.HTTPError as err:
-            if 400 <= err.code <= 499:
-                _log.warning("URL %s was not found (HTTP response code %s), not trying again" % (url, err.code))
+        except HTTPError as err:
+            if not HAVE_REQUESTS:
+                status_code = err.code
+            if 400 <= status_code <= 499:
```

Comment:
Isn't `status_code` potentially undefined here?

Reply:
No, for `HAVE_REQUESTS` because status_code is set as
                `status_code = url_req.status_code`
before
                `url_req.raise_for_status()`